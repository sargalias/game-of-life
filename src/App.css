@import url("https://fonts.googleapis.com/css?family=Raleway");
* {
  margin: 0;
  padding: 0; }

body {
  font-family: 'Raleway', sans-serif;
  -webkit-font-feature-settings: "lnum";
  -moz-font-feature-settings: "lnum";
  font-variant-numeric: normal;
  overflow: hidden; }

.container {
  max-width: 1000px;
  margin-left: auto;
  margin-right: auto;
  margin-top: 3rem;
  text-align: center;
  padding: 0 1rem;
  overflow: hidden; }

#title {
  text-align: center;
  margin-bottom: 1.5rem;
  font-weight: 400; }

.flex-center {
  display: flex;
  align-items: center;
  justify-content: center; }

.button-container {
  margin-bottom: 1.5rem; }
  .button-container button {
    background-color: unset;
    border: 1px solid #ffcab6;
    border-radius: 1rem;
    box-sizing: border-box;
    font-size: 1.3rem;
    margin-left: 0.5rem;
    margin-right: 0.5rem;
    padding: 0.5rem;
    transition: 0.1s; }
    @media screen and (min-width: 500px) {
      .button-container button {
        margin-left: 1rem;
        margin-right: 1rem;
        padding: 1rem; } }
    .button-container button:hover {
      border: 1px solid #ffa483;
      color: coral;
      cursor: pointer; }

#generation {
  font-size: 1.3rem;
  color: coral;
  margin-bottom: 1.5rem; }

#canvas-container {
  overflow: hidden;
  margin-bottom: 1.5rem; }

#attribution p {
  font-size: 1.3rem;
  margin-bottom: 1.5rem; }
  #attribution p#code-repository {
    font-size: 1rem; }
  #attribution p a {
    text-decoration: none;
    color: coral; }

.board {
  margin-bottom: 1.5rem; }
  .board .row:first-child .cell {
    border-top: 1px solid black; }

.row {
  display: flex;
  align-items: center;
  justify-content: center; }
  .row :first-child .cell {
    border-top: 1px solid black; }

.cell {
  width: 12px;
  height: 12px;
  flex: 0 1 auto;
  border-right: 1px solid black;
  border-bottom: 1px solid black;
  display: inline-block;
  box-sizing: border-box; }
  .cell:first-child {
    border-left: 1px solid black; }

.dead {
  background-color: transparent; }

.alive {
  background-color: coral; }

.color-secondary {
  background-color: #cccccc; }
