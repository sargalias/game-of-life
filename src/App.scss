@import url('https://fonts.googleapis.com/css?family=Lato');
@import './input-range.scss';


$color-primary: coral;
$color-secondary: #cccccc;

$font-size-md: 1.3rem;
$size-lg: 1.5rem;
$size-md: 1rem;
$size-sm: 0.5rem;

* {
  margin: 0;
  padding: 0;
}

body {
  font-family: 'Lato', sans-serif;
}

.container {
  max-width: 1000px;
  margin-left: auto;
  margin-right: auto;
  margin-top: $size-lg * 2;
  text-align: center;
  padding: 0 $size-md;
  overflow: hidden;
}

#title {
  text-align: center;
  margin-bottom: $size-lg;
  font-weight: 400;
}

.flex-center {
  display: flex;
  align-items: center;
  justify-content: center;
}

.button-container {
  margin-bottom: $size-lg;
  button {
    background-color: unset;
    border: 1px solid lighten($color-primary, 20%);
    border-radius: 1rem;
    box-sizing: border-box;
    font-size: $font-size-md;
    margin-left: $size-sm;
    margin-right: $size-sm;
    padding: $size-sm;
    transition: 0.1s;
    @media screen and (min-width: 500px) {
      margin-left: 1rem;
      margin-right: 1rem;
      padding: 1rem;

    }
    &:hover {
      border: 1px solid lighten($color-primary, 10%);
      color: $color-primary;
      cursor: pointer;
    }
  }
}

#generation {
  font-size: $font-size-md;
  color: $color-primary;
  margin-bottom: $size-lg;
}

#canvas-container {
  overflow: hidden;
  margin-bottom: $size-lg;
}

#attribution {
  p {
    font-size: $font-size-md;
    margin-bottom: $size-lg;
    &#code-repository {
      font-size: 1rem;
    }
    a {
      text-decoration: none;
      color: $color-primary;
    }
  }
}

.board {
  margin-bottom: $size-lg;
  .row:first-child .cell {
    border-top: 1px solid black;
  }
}

.row {
  display: flex;
  align-items: center;
  justify-content: center;
  :first-child .cell {
    border-top: 1px solid black;
  }
}

.cell {
  width: 12px;
  height: 12px;
  flex: 0 1 auto;
  border-right: 1px solid black;
  border-bottom: 1px solid black;
  display: inline-block;
  box-sizing: border-box;
  &:first-child {
    border-left: 1px solid black;
  }
}

.dead {
  background-color: transparent;
}

.alive {
  background-color: $color-primary;
}

.slider {
  margin-bottom: $size-lg;
}

.slider__title {
  font-size: $font-size-md;
}

.slider__text {
  //border: 1px solid black;
  //padding: $size-sm / 2;
  //border-radius: $size-sm;
  //background-color: $color-primary;
  //color: white;
  font-size: $font-size-md * 0.8;
}

.pattern-picker-container {
  margin-bottom: $size-lg;
}

.pattern-picker-text {
  font-size: $font-size-md;
}

#pattern {
  font-family: 'Lato', sans-serif;
}

input[type=range] {
  width: 100px;
  margin-left: $size-md;
  margin-right: $size-md;
}
