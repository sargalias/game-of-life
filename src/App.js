import React, { Component } from 'react';
import Container from './components/Container';
import Title from './components/Title';
import Generation from './components/Generation';
import Buttons from './components/Buttons';
import Attribution from './components/Attribution';
import Board from './components/Board';
import {generateRandomBoard} from './logic/generateBoard';
import rules from './logic/rules';

import './App.css';


class App extends Component {
  state = {
    generation: 0,
    boardData: [],
    timer: null,
    isRunning: this.props.running
  };

  componentDidMount() {
    this._initialize();
  }

  update = () => {
    const newBoardData = rules(this.state.boardData);
    this.setState((prevState) => ({
      boardData: newBoardData,
      generation: prevState.generation + 1
    }));
  };

  run = () => {
    this.setState((prevState) => {
      if (prevState.timer)
        return {};
      const timer = setInterval(this.update, this.props.interval);
      return {
        timer,
        isRunning: true
      }
    });
  };

  pause = () => {
    clearInterval(this.state.timer);
    this.setState(() => ({
      timer: null,
      isRunning: false
    }));
  };

  clear = () => {
    this.pause();
    this.setState(() => ({
      boardData: generateRandomBoard(this.props.rows, this.props.cols, 0),
      generation: 0
    }));
  };

  _initialize = () => {
    const boardData = generateRandomBoard(this.props.rows, this.props.cols, this.props.chance);
    this.setState(() => ({
      boardData
    }));
    if (this.props.running)
      this.run();
  };

  reset = () => {
    this.clear();
    this._initialize();
  };

  componentWillUnmount() {
    console.log('component will unmount');
    this.pause();
  }

  render() {
    return (
      <Container>
        <Title title="Game Of Life" />
        <Generation generation={this.state.generation} />
        <Buttons onRunClick={this.run} onPauseClick={this.pause} onClearClick={this.clear} onResetClick={this.reset} />
        <Board boardData={this.state.boardData} />
        <Attribution
          authorName="Spyros Argalias"
          authorUrl="https://sargalias.com"
          codeRepositoryUrl="https://github.com/sargalias/game-of-life"
        />
      </Container>
    );
  }
}


export default App;
