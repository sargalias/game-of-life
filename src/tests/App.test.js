import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';


test('app renders correctly', () => {
  const wrapper = shallow(<App chance={0} rows={30} cols={50} running={false} />);
  expect(wrapper).toMatchSnapshot();
});

test('app state.boardData is correct upon mounting', () => {
  const rows = 10;
  const cols = 20;
  const wrapper = shallow(<App chance={0.3} rows={rows} cols={cols} running={false} />);
  const boardData = wrapper.state('boardData');
  expect(boardData.length).toBe(rows);
  expect(boardData[0].length).toBe(cols);
});

test('app state.timer is correct upon mounting', () => {
  const rows = 10;
  const cols = 20;
  const wrapper = shallow(<App chance={0.3} rows={rows} cols={cols} running={false} />);
  const timer = wrapper.state('timer');
  expect(timer).toBe(null);
});

test('app state.generation is correct upon mounting', () => {
  const rows = 10;
  const cols = 20;
  const wrapper = shallow(<App chance={0.3} rows={rows} cols={cols} running={false} />);
  const generation = wrapper.state('generation');
  expect(generation).toBe(0);
});

test('app state.isRunning is correct upon mounting', () => {
  const wrapper1 = shallow(<App chance={0} rows={10} cols={10} running={false} />);
  expect(wrapper1.state('isRunning')).toBe(false);

  const wrapper2 = shallow(<App chance={0} rows={10} cols={10} running={true} />);
  expect(wrapper2.state('isRunning')).toBe(true);
});

test('update works correctly', () => {
  const rows = 5;
  const cols = 3;
  const expectedBoardData = [
    [1, 0, 1],
    [0, 0, 0],
    [0, 0, 0],
    [0, 0, 0],
    [1, 0, 1]
  ];
  const wrapper = shallow(<App chance={1} rows={rows} cols={cols} running={false} />);
  wrapper.instance().update();
  const state = wrapper.state();
  expect(state.generation).toBe(1);
  expect(state.boardData).toEqual(expectedBoardData);
});

test('run works correctly with no current timer', () => {
  const wrapper = shallow(<App chance={1} rows={10} cols={10} running={false} />);
  expect(wrapper.state('timer')).toBe(null);
  expect(wrapper.state('isRunning')).toBe(false);
  wrapper.instance().run();
  expect(wrapper.state('timer')).toBeTruthy();
  expect(wrapper.state('isRunning')).toBe(true);
});

test('run has no effect with already existing timer', () => {
  const wrapper = shallow(<App chance={1} rows={10} cols={10} running={true} />);
  const currentTimer = wrapper.state('timer');
  expect(currentTimer).toBeTruthy();
  wrapper.instance().run();
  expect(wrapper.state('timer')).toBe(currentTimer);
});

test('pause works correctly', () => {
  const wrapper = shallow(<App chance={1} rows={10} cols={10} running={true} />);
  expect(wrapper.state('timer')).toBeTruthy();
  expect(wrapper.state('isRunning')).toBe(true);
  wrapper.instance().pause();
  expect(wrapper.state('timer')).toBe(null);
  expect(wrapper.state('isRunning')).toBe(false);
});

test('clear works correctly', () => {
  const wrapper = shallow(<App chance={0} rows={3} cols={3} running={true} />);
  wrapper.setState({
    boardData: [
      [1, 1, 0],
      [1, 1, 0],
      [0, 0, 0]
    ]
  });
  expect(wrapper.state('timer')).toBeTruthy();

  wrapper.instance().clear();
  expect(wrapper.state('timer')).toBe(null);
  expect(wrapper.state('boardData')).toEqual([
    [0, 0, 0],
    [0, 0, 0],
    [0, 0, 0]
  ]);
  expect(wrapper.state('generation')).toBe(0);
});

test('reset works correctly', () => {
  const wrapper = shallow(<App chance={1} rows={2} cols={2} running={false} />);
  wrapper.setState({
    boardData: [],
    generation: 0,
    timer: null,
    isRunning: null
  });
  expect(wrapper.state('boardData')).toEqual([]);
  wrapper.instance().reset();
  const state = wrapper.state();
  expect(state.boardData).toEqual([[1, 1], [1, 1]]);
  expect(state.generation).toBe(0);
});
