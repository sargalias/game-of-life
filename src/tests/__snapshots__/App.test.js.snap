// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app renders correctly 1`] = `
<Container>
  <Title
    title="Game Of Life"
  />
  <Generation
    generation={0}
  />
  <Buttons
    onClearClick={[Function]}
    onPauseClick={[Function]}
    onResetClick={[Function]}
    onRunClick={[Function]}
  />
  <Board
    boardData={
      Array [
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
      ]
    }
  />
  <Attribution
    authorName="Spyros Argalias"
    authorUrl="https://sargalias.com"
    codeRepositoryUrl="https://github.com/sargalias/game-of-life"
  />
</Container>
`;
